srml/randomness/src/lib.rs

```diff
@@ -0,0 +1,304 @@
+//! 			Ok(())
+//! 		}
+//! 	}
+//! }
+//! # fn main() { }
+//! ```
+
+#![cfg_attr(not(feature = "std"), no_std)]
+
+use rstd::{prelude::*, convert::TryInto};
+use sr_primitives::traits::Hash;
+use support::{decl_module, decl_storage};
+use safe_mix::TripletMix;
+use codec::Encode;
+use system::Trait;
+
+
+fn block_number_to_index<T: Trait>(block_number: T::BlockNumber) -> usize {
+	// '- 2' because on_initialize is called on the first block after genesis - block 2
+	let index = (block_number - 2.into()) % 81.into();
+	index.try_into().ok().expect("something % 81 should always be less than usize")
+}
+
+decl_module! {
+	pub struct Module<T: Trait> for enum Call where origin: T::Origin {
+		fn on_initialize(block_number: T::BlockNumber) {
+			let parent_hash = <system::Module<T>>::parent_hash();
+
+			<RandomMaterial<T>>::mutate(|ref mut values| if values.len() < 81 {
+				values.push(parent_hash)
+			} else {
+				let index = block_number_to_index::<T>(block_number);
+				values[index] = parent_hash;
+			});
+		}
+	}
+}
+
+decl_storage! {
+	trait Store for Module<T: Trait> as System {
+		/// Series of block headers from the last 81 blocks that acts as random seed material. This
+		/// is arranged as a ring buffer with `block_number % 81` being the index into the `Vec` of
+		/// the oldest hash.
+		RandomMaterial get(random_material): Vec<T::Hash>;
+	}
+}
+
+impl<T: Trait> Module<T> {
+	/// Get the basic random seed.
+	///
+	/// In general you won't want to use this, but rather `Self::random` which
+	/// allows you to give a subject for the random result and whose value will
+	/// be independently low-influence random from any other such seeds.
+	pub fn random_seed() -> T::Hash {
+		Self::random(&[][..])
+	}
+
+	/// Get a low-influence "random" value.
+	///
+	/// Being a deterministic block chain, real randomness is difficult to come
+	/// by. This gives you something that approximates it. `subject` is a
+	/// context identifier and allows you to get a different result to other
+	/// callers of this function; use it like `random(&b"my context"[..])`.
+	///
+	/// This is initially implemented through a low-influence "triplet mix"
+	/// convolution of previous block hash values. In the future it will be
+	/// generated from a secure verifiable random function (VRF).
+	///
+	/// ### Security Notes
+	///
+	/// This randomness uses a low-influence function, drawing upon the block
+	/// hashes from the previous 81 blocks. Its result for any given subject
+	/// will be known in advance by the block producer of this block (and,
+	/// indeed, anyone who knows the block's `parent_hash`). However, it is
+	/// mostly impossible for the producer of this block *alone* to influence
+	/// the value of this hash. A sizable minority of dishonest and coordinating
+	/// block producers would be required in order to affect this value. If that
+	/// is an insufficient security guarantee then two things can be used to
+	/// improve this randomness:
```

Comment:
".. will be known far in advance by anyone observing the chain."

Replace "However, ..." with "Any block producer has significant influence over their block hashes bounded only by their computational resources.  Our low-influence function reduces the actual block producer's influence over the randomness, but increases the influence of small colluding groups of recent block producers."

"Some BABE blocks have VRF outputs where the block producer has exactly one bit of influence, either they make the block or they do not make the block and thus someone else makes the next block.  Yet, this randomness is not fresh in all BABE blocks."